syne_tune/stopping_criterions/automatic_termination_criterion.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+import logging
+import numpy as np
+import torch
+from typing import List, Dict, Any
+
+from syne_tune.tuning_status import TuningStatus
+from syne_tune.optimizer.schedulers.searchers.utils import (
+    make_hyperparameter_ranges,
+)
+from torch import Tensor
+from botorch.models import SingleTaskGP
+from botorch.fit import fit_gpytorch_mll
+from botorch.utils.transforms import normalize
+from botorch.acquisition import UpperConfidenceBound
+from botorch.optim import optimize_acqf
+from gpytorch.mlls import ExactMarginalLogLikelihood
+
+logger = logging.getLogger(__name__)
+
+EPSILON_LOG = 1e-10
+NOISE_LEVEL = 1e-3
+
+
+class AutomaticTerminationCriterion(object):
+    """
+    Implements the automatic termination criterion proposed by Makarova et al. The idea is to automatically
+    stop the optimization process if the expected gain in regret is smaller than the observation noise
+    of the objective.
+
+    Automatic Termination for Hyperparameter Optimization
+    Anastasia Makarova and Huibin Shen and Valerio Perrone and Aaron Klein and Jean Baptiste Faddoul and Andreas Krause
+    and Matthias Seeger and Cedric Archambeau
+    First Conference on Automated Machine Learning (Main Track)
+    2022
+
+    :param config_space: Configuration space for evaluation function
+    :param metric: The metric to be monitored.
+    :param mode: The mode to select the top results ("min" or "max")
+    :param threshold: The threshold on the regret. We stop the optimization process when it's unlikely that the
+    expected decrease in regret compared to the global optimum is higher than this threshold.
+    :param beta: Multiplier on the standard deviation to compute the upper and lower confidence bound.
+    :param seed: Seed for the random numer generator.
+    :param warm_up: Defines the minimum umber of data points before we start fitting the GP.
+    :param topq: We consider only the top q-% data points to fit the GP.
+    """
+
+    def __init__(
+        self,
+        config_space: Dict[str, Any],
+        metric: str,
+        threshold: float,
+        mode: str = "min",
+        beta: float = 1.0,
+        seed: int = None,
+        warm_up: int = 10,
+        topq: float = 0.5,
```

Comment:
You should probably assert that the value in ]0,1[, the user may pass you an integer for instance by mistake.